viewing_party/party.py

```diff
@@ -0,0 +1,193 @@
+    has watched is empty. 
+    """
+    genre_frequency = {}
+    if len(user_data["watched"]) == 0:
+        return None
+    for movie in user_data["watched"]:
+        genre = movie["genre"]
+        if genre not in genre_frequency:
+            genre_frequency[genre] = 1
+        else:
+            genre_frequency[genre] += 1
+
+    for genre, frequency in genre_frequency.items():
+        if frequency  == max(genre_frequency.values()):
+            return genre
+
+# # Wave 3
+# Helper function that will be used for waves 3-5
+def get_friends_watched(user_data):
+    """
+    Depending on the parameter, the function returns
+    different types of information about the movies that the
+    user's friends have watched. 
+    """
+    friends_watched = []
+    for watched_movies in user_data["friends"]:
+        for title in watched_movies["watched"]:
+            friends_watched.append(title)
+    return friends_watched
+
+def get_unique_watched(user_data):
+    """
+    This function returns a list of dictionaries that represents a 
+    list of movies that the user has watched, but none of their 
+    friends have watched. 
+    """
+    friends_watched = get_friends_watched(user_data)
+
+    unique_watched = []
+    for title in user_data["watched"]:
+        if title not in friends_watched:
+            if title not in unique_watched:
+                unique_watched.append(title)
+    return unique_watched
+
+def get_friends_unique_watched(user_data):
+    """
+    This function returns a list of dictionaries that represents 
+    a list of movies that the user's friends have watched, but the 
+    user has not watched. 
+    """
+    friends_watched = get_friends_watched(user_data)
+
+    friends_unique_watched = []
+    for title in friends_watched:
+        if title not in user_data["watched"]:
+            if title not in friends_unique_watched:
+                friends_unique_watched.append(title)
+    return friends_unique_watched
+
+# Wave 4
+def get_available_recs(user_data):
+    """
+    This function returns a list of recommended movies that 
+    the user has not watched but one of the user's friends has 
+    watched. This list of recommended movies is only hosted by 
+    a service that is in the user's subscriptions.
+    """
+    friends_watched = get_friends_watched(user_data)
+    
+    subscriptions_match = []
+    for title_host in friends_watched:
+        if title_host["host"] in user_data["subscriptions"]:
+            if title_host not in subscriptions_match: 
+                subscriptions_match.append(title_host)
+    
+    watched_titles_only = [movie["title"] for movie in user_data["watched"]]
+
+    recommendations = []
+    if not user_data["watched"]: 
+        return subscriptions_match
+    for title_host in subscriptions_match:
+        if title_host["title"] not in watched_titles_only:
+            recommendations.append(title_host)
+    return recommendations
```

Comment:
this looks very nice! breaking things up so you understand what is happening is always good! I wonder, though, if we could combine some of these for loops  and create a larger compound conditional hmmmm